### ISSUE TITLE
Implement a way to get outputs from `execute`

### ISSUE THREAD

**Issue by tecracer-theinen** [chef/chef#9131]
### Describe the Enhancement:
In various customer projects I have seen the need to execute commands and work with their return values. Sometimes it is an API token, sometimes the return value of a binary which is later used in curl requests.

### Describe the Need:
Users who need to access the output from other tools or remote systems (in Target Mode).

### Current Alternative:
Writing custom resources which use `shell_out!` or writing to a file which is read later on.

### Can We Help You Implement This?:
The main discussion point is probably where to export the data to, the existing Chef AWS cookbook's `ssm_parameter_store` resource uses `node.run_state` as a temporary scratch space.

From then, adding an `exports` property to the `execute` resource and parsing stdout via regex is pretty easy.

By using `export`, the functionality should be inherited from all resources derived from it.

The functionality should probably be seen as being on an advanced level, because it needs understanding of some desired state functionality (if the execute resource is not run as it has a guard applied, the output will not exist).

I am using an implementation of this in some projects already and would love to make a PR if the functionality is wanted and using `node.run_state` is accepted. Feedback appreciated

**Comment by lamont-granquist** [comment]
The `ssm_parameter_store` should not have used the node.run_state as scratch space.  I am 1000% opposed to extending that bad pattern, and I probably shouldn't have allowed that PR to go in without changing that.

You also need to include a more compelling use case here, because if you're just trying to run a command and parse the output of it, then you absolutely should never use an execute resource and should use `shell_out`.

If you look at how every single package resource is written they use shell_out's to grab various bits of data about what packages are available or installed or whatever.  And prior to unified_mode being a thing it would have been awful to even consider rewriting those as execute resources by mixing up compile+converge mode.  So my gut is that this is a philosophical issue where users believe that they've heard that they need to use chef resources everywhere, when actually they need to learn when chef resources are appropriate and when they are not.

I'm not entirely opposed to pulling information out of resources, but it needs a much better use case that isn't one where its obvious that trying to pull information out of resources is clearly the poorer choice than just doing the programming imperatively.  The resource should really be doing something that changes the running state of the system (and scribbling on a private file in the file_cache_path or tempfile or something like that is NOT something that I consider to be the "running state of the system" when its just an intermediate state of remote information which has been pulled locally for data massaging.  It needs to be something which is consumed by the O/S or applications on the system, if it is just state that chef creates so that chef consumes it, then it is not the running state.

**Comment by lamont-granquist** [comment]
(the http_resource is another awful example -- the right answer there is to use Chef::HTTP::Simple directly in "imperative mode" just like using shell_out -- what might made that experience better is just a helper around that injected as a method into the DSL)

**Comment by tas50** [comment]
@lamont-granquist Is going to leave some additional thoughts here from our triage today

**Comment by lamont-granquist** [comment]
Yeah I think I've relented, but the correct way of doing this is using a `state_property` on the new_resource which is loaded by the provider.  It is possible that we need another kind of property to document this stuff correctly while making it clear that these are not to be set, but are consumed.  Now that we have unified_mode this would work reasonably well, before it would be horrible.
